frontend/actions.js

```diff
@@ -98,58 +89,149 @@ export default ({setState, getState}) => {
     displayAboutOverlay: !state.displayAboutOverlay,
   })
 
-  const inputAddress = (state, e) => ({
-    sendAddress: e.target.value,
+  const clearSendAmountValidation = (state) => ({
+    sendAmount: Object.assign(state.sendAmount, {validationMsg: ''}),
+  })
+
+  const clearSendAddressValidation = (state) => ({
+    sendAddress: Object.assign(state.sendAddress, {validationMsg: ''}),
   })
 
-  // is being called debounced, thus we need to getState
-  // TODO maybe it'd be good practice to only use getState and avoid this problem?
+  const validateSendAddress = (state) => {
+    if (!Cardano.isValidAddress(state.sendAddress.fieldValue)) {
+      setState({
+        sendAddress: Object.assign(state.sendAddress, {validationMsg: 'Invalid address'}),
+      })
+      return false
+    }
+    return true
+  }
+
+  const validateSendAmount = (state) => {
+    if (isNaN(parseFloat(state.sendAmount.fieldValue))) {
+      setState({
+        sendAmount: Object.assign(state.sendAmount, {
+          validationMsg: 'Invalid format: Amount has to be a number',
+        }),
+      })
+      return false
+    }
+    if (parseFloat(state.sendAmount.fieldValue).toFixed(6) <= 0) {
+      setState({
+        sendAmount: Object.assign(state.sendAmount, {
+          validationMsg: 'Invalid format: Amount has to be a positive number',
+        }),
+      })
+      return false
+    }
+    if (parseFloat(state.sendAmount.fieldValue).toFixed(6) > state.balance / 1000000) {
+      setState({
+        sendAmount: Object.assign(state.sendAmount, {
+          validationMsg: `Insufficient funds. Your balance is ${state.balance / 1000000} ADA.`,
+        }),
+      })
+      return false
+    }
+    return true
+  }
+
+  const validateFee = (state) => {
+    if (parseFloat(state.transactionFee) < 0) {
```

Comment:
wat ? this can happen ?

Reply:
The Cardano library function for fee computation returns a negative value (-1 * computed fee) if the fee cannot be applied (i.e. you don't have enough funds). Subsequently, this value can be used to estimate (but quite exactly for most practical purposes) the maximum sendable amount (you substract the fee from your balance).

Yes, it would probably be cleaner to have two variants of the function, one returning the computed fee no matter what and the other one throwing an exception if the fee cannot be applied, or simply to always return the computed fee and let the caller of the function do the math themself.